### PR TITLE
Better error messages

### PR DESCRIPTION
**Pull request by robertjd**
Refactoring the internal request executor to provide better error messages, enabling developers to troubleshoot the problem that an end-user may have.

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -43,10 +43,10 @@ module.exports = function(config) {
     browsers: ['Chrome'],
 
     browserify: {
-      watch: true
+      watch: true,
+      debug: true
```
Comment (timothyej) [inline karma.conf.js]: Did you intend to commit this (`debug: true`)?

File: lib/client.js [hunk 2]
```diff
@@ -1,47 +1,104 @@
...
+/**
+ * Creates a Stormpath.js Client
+ *
+ * A client is meant to encapsulate the communication
+ * with Stormpath's REST API.
+ *
+ * @constructor
+ * @param {object} options configuration options
+ * @param {function} readyCallback called when the client has
+ * initialized with it's needed data from the REST API.
```
Comment (timothyej) [inline lib/client.js]: Typo: `it's` => `its`

File: lib/client.js [hunk 3]
```diff
@@ -1,47 +1,104 @@
...
+      if (err) {
+        if (err.status === 401) {
+          return cb(new Error(strings.errors.INITIAL_JWT_REJECTED));
+        }
+        return cb(err);
+      }
+
+      /*
+        Assert that the response got a new auth token header.  If it did not,
+        ther is likely a proxy or firewall that is stripping it from the
```
Comment (timothyej) [inline lib/client.js]: Typo: `ther` => `there`

File: lib/client.js [hunk 4]
```diff
@@ -51,97 +108,178 @@ Client.prototype.getCachedOrganizationNameKey = function() {
...
     },
     callback || utils.noop
   );
 
 };
 
-Client.prototype.register = function register(data,callback) {
-  if(typeof data!=='object'){
+/**
+ * Make an account creation atempt against the REST API, given the input data
```
Comment (timothyej) [inline lib/client.js]: Type: `atempt` => `attempt`

File: lib/client.js [hunk 5]
```diff
@@ -51,97 +108,178 @@ Client.prototype.getCachedOrganizationNameKey = function() {
...
+ * }
+ * </pre>
+ * @param  {Function} callback - If no error is given, the password was reset
+ * successfully.  If an error, the password strength validation error will be
+ * provided.
+ */
+
+Client.prototype.setAccountPassword = function setAccountPassword (passwordVerificationTokenResource,newPassword,callback) {
+
+  var client = this;
```
Comment (timothyej) [inline lib/client.js]: +1 for using a better name than `self` :)

File: lib/defer-callback.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * is done processing (aka Angular's digest scheme)
+ *
+ * @function
+ * @param  {function} cb - The callback to call at a later time
+ * @return {array} - args - The array of arguments to apply to the callback
+ */
+function deferCallback (cb,args) {
+  setTimeout(function () {
+    cb.apply(null,args);
+  },1);
```
Comment (timothyej) [inline lib/defer-callback.js]: I usually use `0` instead of `1` :)

File: lib/idsite-request-executor.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  requestorCallback(null,body);
+
+};
+
+/**
+ * Makes a request of the REST API, given the options for the xhr library.
+ * The cached auth token is used for authentication with the REST API.
+ *
+ * @param  {object}   xhrRequestOptions - Options for the call to `xhr(options)`
+ * @param  {Function} callback - The callback to cal lwith the result, as decoraged
```
Comment (timothyej) [inline lib/idsite-request-executor.js]: Typos:

* `cal lwith` => `call with`
* `decoraged` => `decorated`

File: lib/strings.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "errors": {
+    "JWT_NOT_FOUND": "JWT not found as url query parameter.",
+    "NOT_A_JWT": "JWT does not appear to be a property formatted JWT.",
+    "MALFORMED_JWT_CLAIMS": "The JWT claims section is malfomed and could not be decoded as JSON.",
+    "NO_AUTH_TOKEN_HEADER": "HTTP response does not contain Authorization header",
```
Comment (timothyej) [inline lib/strings.json]: These three lines does not end with a punctation, but the first three does.

File: lib/strings.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+    "INVALID_AUTH_TOKEN_HEADER": "HTTP response has an invalid Authorization header",
+    "INITIAL_JWT_REJECTED": "The JWT used to initialized the client was rejected"
```
Comment (timothyej) [inline lib/strings.json]: Typo: `initialized` => `initialize`

**Comment by timothyej** [comment]
Other than my comments this looks good :+1:

File: karma.conf.js (hunk 1 above)
Comment (robertjd) [inline karma.conf.js]: Yes, it shows the line numbers correctly when showing errors in tests.

File: lib/client.js (hunk 5 above)
Comment (robertjd) [inline lib/client.js]: :)

**Comment by robertjd** [comment]
Thanks @timothyej for the review!  I updated accordingly, merging.

**Pull request merged**